Normalize package.json

jlord opened the pull request:
In an previously merged PR #6582 we just normalized the repository url. Here we normalize the entire `package.json` in the same way that npm does. 

TODO:
- [x] Fix Atom bundled packages with bad package data in order for this to go :green_apple: 

Closes https://github.com/atom/atom/issues/6601

jlord committed:
normalize all of package.json

use different module, catch throws, and still remove the ‘git’ prefix
from urls.

kevinsawicki commented:
It would be great to know how much time this takes for the bundled core packages just to make sure it doesn't impact startup time in a noticeable way.

jlord commented:
> know how much time this takes

@kevinsawicki :+1: definitely.

Also, should mentioned I started the PR early cause tests were being weird on my machine and I wanted to see what happened to them here.

jlord committed:
Don't need to replace now, use module directly

jlord committed:
Merge branch 'master' into jl-all-your-normalizes

jlord committed:
merge

jlord committed:
Add error message to Jasmine output

jlord committed:
Normalize bundled package data

Warn if contains incomplete package.json data

jlord committed:
Only normalize unpublished packages

Packages published to APM are normalized already, this catches locally
linked/developed packages.

jlord committed:
Fail when bundled packages have bad metadata

jlord commented:
Update:

We realized that published packages are already getting the benefit of `normalized-package-data`. In order to save on :watch: we're only going to normalize package which haven't been published.

We're also checking Atom's bundled packages as well and failing the build if they are missing something. This build will therefore fail until we add licenses to:

- [ ] archive-view
- [ ] autocomplete
- [ ] language-gfm
- [ ] language-toml
- [ ] spell-check

jlord committed:
Merge branch 'master' into jl-all-your-normalizes

jlord committed:
Merge branch 'master' into jl-all-your-normalizes

jlord committed:
Remove 'git+' prefix

jlord committed:
Only remove 'git+' if 'git' type matches

jlord committed:
Merge branch 'master' into jl-all-your-normalizes

jlord committed:
Remove `?`

jlord committed:
No spaces in package.json name

jlord committed:
Check for repository, log error (temp)

jlord committed:
Check for repository data

jlord committed:
Fixture name to match fixture dir

jlord commented:
It is green—it almost seems too good to be true. I guess it was the `?` repository check? Though I'm not 100% sure why. @kevinsawicki would you mind giving it another :eyes: ?

kevinsawicki commented on src/package.coffee:
```diff
@@ -26,16 +26,12 @@ class Package
+      normalizePackageData ?= require 'normalize-package-data'
+      normalizePackageData(metadata)
+      if metadata.repository?.type is 'git'
+        metadata.repository.url = metadata.repository.url.replace(/^git\+/, '')
```
I think you may want to check `metadata.repository.url` before calling `replace` on it. I think it could be undefined even with `repository.type` is `'git'`.

```sh
> metadata = {"name": "hey", "version": "1.0.0", "repository": {"type": "git"}}
{ name: 'hey', version: '1.0.0', repository: { type: 'git' } }
> normalize = require('normalize-package-data')
[Function: normalize]
> normalize(metadata)
undefined
> metadata
{ name: 'hey',
  version: '1.0.0',
  repository: { type: 'git' },
  readme: 'ERROR: No README data found!',
  _id: 'hey@1.0.0' }
> metadata.repository.url
undefined
>
```

jlord replied:
:+1:

kevinsawicki commented on src/package.coffee:
```diff
@@ -45,12 +41,13 @@ class Package
           metadata = CSON.readFileSync(metadataPath)
+          @normalizeMetadata(metadata)
         catch error
+          console.log("Thew Error", metadata.name, error)
```
:fire:

jlord replied:
Gah! I read over the diff twice. :fire: ing.

50Wliu commented on spec/fixtures/packages/package-with-empty-activation-commands/package.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "no events",
+  "name": "package-with-consumed-services",
```
`package-with-empty-activation-commands`?

kevinsawicki replied:
:+1:

jlord replied:
:blush:

jlord committed:
Correct package name, remove log

jlord committed:
Check for url before replacing

jlord committed:
merge

jlord committed:
Bump version of autocomplete-plus@2.16.0

Add license filed for passing tests and standardization.

jlord committed:
Merge master

kevinsawicki committed:
Merge branch 'master' into jl-all-your-normalizes

kevinsawicki committed:
Remove metadata return

kevinsawicki committed:
Check that url is string before replacing prefix

kevinsawicki commented on src/package.coffee:
```diff
@@ -26,16 +26,11 @@ class Package
+    unless metadata?._id
+      normalizePackageData ?= require 'normalize-package-data'
+      normalizePackageData(metadata)
+      if metadata.repository?.type is 'git' and typeof metadata.repository.url is 'string'
```
@jlord after reviewing this again, I thought it would be better to tighten up this check to only call `replace` if the url is  a `string`.

That way in the weird case where they have a number of boolean value there, it won't blow up with a `undefined is not a function` errors.

kevinsawicki commented:
:ship: :sunglasses:
